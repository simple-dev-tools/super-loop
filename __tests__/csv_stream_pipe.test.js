const csv = require('csv-parser');
const { Writable } = require('stream')
const fs = require('fs');
const SuperLoop = require('../src/superloop');

describe('SuperLoop custom Readable as stream', () => {

    it('#process csv file', async () => {

        const upstream = fs.createReadStream(`${__dirname}/small_data.csv`).pipe(csv())

        const loop = new SuperLoop();
        let counter = 0;
        const f = () => {
            counter += 1;
        }

        await loop.pipeFrom(upstream).invoke(f).exec();
        expect(counter).toBe(3);

    });

    it('#process csv file and write to downstream', async () => {

        const upstream = fs.createReadStream(`${__dirname}/small_data.csv`).pipe(csv())

        const loop = new SuperLoop();
        let counter = 0;

        const f = (rawEvent) => {
            counter += 1;
            return {
                data: {
                    index: rawEvent.Index,
                    height: rawEvent.Height,
                    weight: rawEvent.Weight
                }
            }
        }


        const wResults = [];
        const writer = new Writable({
            objectMode: true,
            write(data, _, done) {
                wResults.push(data)
                done()
            }
        })

        await loop.pipeFrom(upstream).invoke(f).pipeTo(writer).exec();
        expect(counter).toBe(3);
        expect(wResults).toMatchSnapshot()

    });

})