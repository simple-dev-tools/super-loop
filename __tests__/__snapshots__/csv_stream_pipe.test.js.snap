// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuperLoop custom Readable as stream #process csv file and write to downstream 1`] = `
Array [
  Object {
    "data": Object {
      "height": "65.78331",
      "index": "1",
      "weight": "112.9925",
    },
  },
  Object {
    "data": Object {
      "height": "71.51521",
      "index": "2",
      "weight": "136.4873",
    },
  },
  Object {
    "data": Object {
      "height": "69.39874",
      "index": "3",
      "weight": "153.0269",
    },
  },
]
`;
